extensions/metadata/interface/src/main/java/org/exist/storage/md/CollectionEvents.java

```diff
@@ -29,113 +29,142 @@
+    }
+
+    @Override
+    public void beforeCreateCollection(DBBroker broker, Txn txn, XmldbURI uri) throws TriggerException {
+        // System.out.println("beforeCreateCollection "+uri);
+    }
+
+    @Override
+    public void afterCreateCollection(DBBroker broker, Txn txn, Collection collection) {
+        // System.out.println("afterCreateCollection "+collection.getURI());
+        try {
+            MDStorageManager.get().md.addMetas(collection);
+        } catch (Throwable e) {
+            MDStorageManager.LOG.fatal(e,e);
+        }
+
+//        index(collection);
+    }
+
+    @Override
+    public void beforeCopyCollection(DBBroker broker, Txn txn, Collection collection, XmldbURI newUri) throws TriggerException {
+    }
+
+    @Override
+    public void afterCopyCollection(DBBroker broker, Txn txn, Collection collection, XmldbURI oldUri) {
+        // System.out.println("afterCopyCollection "+collection.getURI());
+        try {
+            MDStorageManager.get().md.copyMetas(oldUri, collection);
+        } catch (Throwable e) {
+            MDStorageManager.LOG.fatal(e,e);
         }
 
-		final XmldbURI uri = collection.getURI();
+//        index(collection);
+    }
 
-		for(Iterator<XmldbURI> i = collection.collectionIterator(broker); i.hasNext(); ) {
-            final XmldbURI childName = i.next();
-            //TODO : resolve URIs !!! name.resolve(childName)
-            final Collection child = broker.openCollection(uri.append(childName), Lock.NO_LOCK);
-            if(child == null) {
-//                LOG.warn("Child collection " + childName + " not found");
-            } else {
+    @Override
+    public void beforeMoveCollection(DBBroker broker, Txn txn, Collection collection, XmldbURI newUri) throws TriggerException {
+        // System.out.println("beforeMoveCollection "+collection.getURI());
+        try {
+            for (Iterator<DocumentImpl> i = collection.iterator(broker); i.hasNext();) {
+                DocumentImpl doc = i.next();
                 try {
-                	deleteCollectionRecursive(broker, child);
-                } finally {
-                    child.release(Lock.NO_LOCK);
+                    MDStorageManager.get().md.moveMetas(collection.getURI().append(doc.getFileURI()), newUri.append(doc.getFileURI()));
+                } catch (Throwable e) {
+                    MDStorageManager.LOG.fatal(e,e);
                 }
             }
+        } catch (PermissionDeniedException e) {
+            throw new TriggerException(e);
+        }
+    }
+
+    @Override
+    public void afterMoveCollection(DBBroker broker, Txn txn, Collection collection, XmldbURI oldUri) {
+        // System.out.println("afterMoveCollection "+oldUri+" to "+collection.getURI());
+        try {
+            MDStorageManager.get().md.moveMetas(oldUri, collection.getURI());
+        } catch (Throwable e) {
+            MDStorageManager.LOG.fatal(e,e);
+        }
+
+        // removeBinary(oldUri);
```

Comment:
@adamretter cleanup a bit, commented out code is features that is not at eXist's develop branch yet